### ISSUE TITLE
Bogus warning "is not needed and will not be emitted"

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#22821]
|  |  |
| --- | --- |
| Bugzilla Link | [22447](https://llvm.org/bz22447) |
| Version | unspecified |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
Given

namespace {
struct foo {
  foo(int = bar) {}
  static int bar;
};
int foo::bar;
}
foo x;

clang warns: "variable 'bar' is not needed and will not be emitted"

but @&#8203;_ZN12_GLOBAL__N_13foo3barE is emitted. Commenting the "int foo::bar"" also produces the warning:

variable '(anonymous namespace)::foo::bar' has internal linkage but is not defined

**Comment by wheatman** [comment]
This seems to have been fixed between 3.8.1 and 3.9.0
https://godbolt.org/z/eW5TE17jq

for reference

code 
```
namespace {
struct foo {
  foo(int = bar) {}
  static int bar;
};
int foo::bar;
}
foo x;
```

old warning with `-Wall`
```
<source>:6:10: warning: variable 'bar' is not needed and will not be emitted [-Wunneeded-internal-declaration]
int foo::bar;
         ^
1 warning generated.
Compiler returned: 0
```
